domain/src/main/java/com/teamagam/gimelgimel/domain/dynamicLayers/OnDynamicLayerListingClickInteractor.java

```diff
@@ -0,0 +1,130 @@
+    return factory.create(Observable.just(mDynamicLayer),
+        dlObservable -> dlObservable.map(this::getNewVisibilityState)
+            .doOnNext(this::setNewVisibility)
+            .filter(isShown -> isShown)
+            .doOnNext(flag -> goToExtent()));
+  }
+
+  private boolean getNewVisibilityState(DynamicLayer dynamicLayer) {
+
+    return !mDynamicLayerVisibilityRepository.isVisible(dynamicLayer.getId());
+  }
+
+  private void setNewVisibility(Boolean newVisibilityState) {
+    mDynamicLayerVisibilityRepository.addChange(
+        new DynamicLayerVisibilityChange(newVisibilityState, mDynamicLayer.getId()));
+  }
+
+  private void goToExtent() {
+    Geometry blockingEnvelope = mDynamicLayerEnvelopeExtractor.extract(mDynamicLayer);
+    mGoToLocationMapInteractorFactory.create(blockingEnvelope).execute();
+  }
+
+  private class DynamicLayerEnvelopeExtractor {
+
+    private ExtractPointsVisitor mPointsExtractorVisitor = new ExtractPointsVisitor();
+
+    public Geometry extract(DynamicLayer dl) {
+      List<PointGeometry> allPoints =
+          getAllPoints(Lists.transform(dl.getEntities(), GeoEntity::getGeometry));
+      return buildEnvelope(allPoints);
+    }
+
+    private List<PointGeometry> getAllPoints(List<Geometry> geometries) {
+      List<PointGeometry> res = new ArrayList<>();
+      for (Geometry g : geometries) {
+        g.accept(mPointsExtractorVisitor);
+        res.addAll(mPointsExtractorVisitor.getPoints());
+      }
+      return res;
+    }
+
+    private Geometry buildEnvelope(List<PointGeometry> allPoints) {
+      List<Double> lats = Lists.transform(allPoints, PointGeometry::getLatitude);
+      List<Double> longs = Lists.transform(allPoints, PointGeometry::getLongitude);
+
+      double minX = Collections.min(lats);
+      double maxX = Collections.max(lats);
+
+      double minY = Collections.min(longs);
+      double maxY = Collections.max(longs);
+
+      return createEnvelope(minX, maxX, minY, maxY);
+    }
+
+    private Geometry createEnvelope(double minX, double maxX, double minY, double maxY) {
+      PointGeometry bottomLeft = new PointGeometry(minX, minY);
+      PointGeometry topLeft = new PointGeometry(minX, maxY);
+      PointGeometry topRight = new PointGeometry(maxX, maxY);
+      PointGeometry bottomRight = new PointGeometry(maxX, minY);
+      return new Polygon(Lists.newArrayList(bottomLeft, topLeft, topRight, bottomRight));
+    }
+
+    private class ExtractPointsVisitor implements GeometryVisitor {
+      List<PointGeometry> mPoints = new ArrayList<>();
+
+      @Override
+      public void visit(PointGeometry point) {
+        mPoints.add(point);
+      }
+
+      @Override
+      public void visit(Polygon polygon) {
+        mPoints = polygon.getPoints();
+      }
+
+      @Override
+      public void visit(Polyline polyline) {
+        mPoints = polyline.getPoints();
+      }
+
+      public List<PointGeometry> getPoints() {
+        return mPoints;
+      }
+    }
+  }
+}
```

Comment:
It's done in an inner class...
Removing it to a a separate class is just semantics...
Given that the file is short, I'd say it'll even reduce the readability